docs/source/tutorial/data-source.md

```diff
@@ -3,6 +3,424 @@ title: "2. Hook up your data sources"
+        name: launch.mission_name,
+        missionPatch: launch.links.mission_patch_small
+      },
+      year: launch.launch_year,
+      rocket: {
+        id: launch.rocket.rocket_id,
+        name: launch.rocket.rocket_name,
+        type: launch.rocket.rocket_type,
+      },
+      launchSuccess: launch.launch_success,
+    };
+  }) : [];
+}
+```
+
+In the code above, `this.get('launches')`, makes a `GET` request to `https://api.spacexdata.com/v2/launches` and stores the returned data in the `res` variable. If the `res` variable is not empty, then the `getAllLaunches` method returns an object that corresponds with the schema fields of the `Launch` schema type, else, an empty array is returned.
+
+Let's refactor the `getAllLaunches` method to be a lot cleaner and concise. Copy the `launchReducer` method below and add to the file. Now, refactor the `getAllLaunches` method to use the `launchReducer` method as shown below:
+
+_src/datasources/launch.js_
+
+```js
+...
+launchReducer(launch) {
+  return {
+    id: launch.flight_number || 0,
+    cursor: `${launch.launch_date_unix}`,
+    mission: {
+      name: launch.mission_name,
+      missionPatch: launch.links.mission_patch_small
+    },
+    year: launch.launch_year,
+    rocket: {
+      id: launch.rocket.rocket_id,
+      name: launch.rocket.rocket_name,
+      type: launch.rocket.rocket_type,
+    },
+    launchSuccess: launch.launch_success,
+  };
+}
+
+async getAllLaunches() {
+  const res = await this.get('launches');
+
+  return res && res.length ? res.map(l => this.launchReducer(l)) : [];
+}
+```
+
+With the above changes, we can easily make changes to the `launchReducer` method while the `getAllLaunches` method stays lean and concise. The `launchReducer` method also makes testing the `LaunchAPI` data source class easier. The later part of this tutorial covers testing!
+
+Next, let's take care of getting a specific launch. Add the method, `getLaunchById`, and `getLaunchesByIds` to the `LaunchAPI` class.
+
+```js
+...
+async getLaunchById({ launchId }) {
+  const res = await this.get('launches', { flight_number: launchId });
+  return this.launchReducer(res[0]);
+}
+
+async getLaunchesByIds({ launchIds }) {
+  return Promise.all(
+    launchIds.map(launchId => this.getLaunchById({ launchId })),
+  );
+}
+```
+
+The `getLaunchById` method takes in a flight number and returns the data for a particular launch, while `getLaunchesByIds` returns several launches based on their respective `launchIds`. `Promise.all()` takes an array of promises and returns a single promise that resolves when all the promises in the array have been resolved with their fulfilled values.
+
+<h2 id="database">Connect a database</h2>
+
+A data store is needed for saving and fetching user information. It's also important for user trips. Let's make use of [SQLite](https://www.sqlite.org) for our app's database. SQLite is a self-contained, light-weight, zero-configuration and embedded SQL database engine.
```

Comment:
We're going to take care of all the Sequelize code for them since we don't want the tutorial to focus too much on this part. Please slim this section down so it just focuses on the Apollo specific parts.